pkg/crd/flatten.go

```diff
@@ -26,47 +27,176 @@ import (
 	"sigs.k8s.io/controller-tools/pkg/loader"
 )
 
+// ErrorRecorder knows how to record errors.  It wraps the part of
+// pkg/loader.Package that we need to record errors in places were it might not
+// make sense to have a loader.Package
+type ErrorRecorder interface {
+	// AddError records that the given error occurred.
+	// See the documentation on loader.Package.AddError for more information.
+	AddError(error)
+}
+
+// isOrNil checks if val is nil if val is of a nillable type, otherwise,
+// it comparse val to valuInt (which should probably be the zero value).
+func isOrNil(val reflect.Value, valInt interface{}, zeroInt interface{}) bool {
+	switch valKind := val.Kind(); valKind {
+	case reflect.Chan, reflect.Func, reflect.Interface, reflect.Map, reflect.Ptr, reflect.Slice:
+		return val.IsNil()
+	default:
+		return valInt == zeroInt
+	}
+}
+
 // flattenAllOfInto copies properties from src to dst, then copies the properties
 // of each item in src's allOf to dst's properties as well.
-func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps) {
+func flattenAllOfInto(dst *apiext.JSONSchemaProps, src apiext.JSONSchemaProps, errRec ErrorRecorder) {
```

Comment:
why not return an aggregated error? No need for a complex ErrorRecorder construct.

Reply:
aggregated errors then need to be extracted later, and this is the pattern used almost literally everywhere else in controller-tools, since it allows us to attach better error data in many cases.